src/main/scala/com/algorithmia/handler/AbstractAlgorithm.scala

```diff
@@ -0,0 +1,27 @@
+package com.algorithmia.handler
+import play.api.libs.json._
+import scala.reflect.runtime.universe._
+
+import scala.util.{Failure, Success, Try}
+
+
+ object AbstractAlgorithm {
+
+   trait AbstractAlgorithm[-I, +O] {
+     def apply(input: I): Try[O]
+
+     def load(): Try[Unit] = Success(())
+   }
+
+   implicit class WeakTypeDetector[I: WeakTypeTag](related: AbstractAlgorithm[I, _]) {
+     def getType: Type = {
+       weakTypeOf[I]
+     }
+   }
+   implicit class TypeDetector[I: TypeTag](related: AbstractAlgorithm[I, _]) {
+     def getType: Type = {
+       typeOf[I]
+     }
+   }
```

Comment:
These seem to ensure that we're always able to get the input type successfully, including if it's a complex type or primitive type. My hunch is, can we reuse this same style of logic to automagically get the Reads/Writes implicit values if they exist, and if not - generate them?